### ISSUE TITLE
Prompts don't initialize attemptCount until the next turn

### ISSUE THREAD

**Issue by v-kydela** [microsoft/botbuilder-js#1458]
## Versions
master branch

## Describe the bug
A prompt state's attempt count is not initialized in [`beginDialog`](https://github.com/microsoft/botbuilder-js/blob/master/libraries/botbuilder-dialogs/src/prompts/prompt.ts#L177-L196) like [it is in .NET](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder.Dialogs/Prompts/Prompt.cs#L98-L101). This is a parity issue because it will result in different behavior. If a prompt's state is accessed from the bot code before [`continueDialog`](https://github.com/microsoft/botbuilder-js/blob/master/libraries/botbuilder-dialogs/src/prompts/prompt.ts#L198-L238) is called the first time then the attempt count will be undefined, whereas it would be 0 in .NET.

## To Reproduce
1. Call a dialog that puts a prompt on top of the stack
2. Manually access `state.state['attemptCount']` from the bot code outside the dialog by getting the active dialog from the stack
3. Observe the value

## Expected behavior
The JS and .NET repos should yield the same value by either both initializing it when beginning the dialog or both not initializing it. I've posted this in the JS repo because my thought is that this is the repo that should change to match the other one.

## Additional context
Related: https://github.com/microsoft/botbuilder-js/issues/1195
Related: https://github.com/microsoft/botbuilder-dotnet/issues/3049

[bug]

**Comment by axelsrz** [comment]
@stevengum could you take a look into this

**Comment by stevengum** [comment]
@v-kydela I don't see this as a large issue, as the end behavior is the same between the two languages. To my knowledge, we haven't promised _state_ parity, only behavior parity.

True, someone manually checking the stack was not an accounted for scenario, but again the end behavior is the same. If it was a larger customer concern (aka common scenario), we would be seeing more reports of this.

I don't see this as critical, we should leave it as is.

@johnataylor can you chime in?

**Comment by stevengum** [comment]
I erroneously did not add my rationale for why we shouldn't seek to address this parity issue. Sorry!

First and foremost, it's a public contract. We can't change the behavior now because customers may be relying on an `undefined` and not a `0`. @v-kydela accessed the state and saw an `undefined`, not a `0` like in C#. To change this value to `0` when bots may expect `undefined` is a breaking change.

It's a parity issue and the code _should have_ been aligned from the start. But now the underlying state is locked in.

Secondly, the behavior that the user sees when interacting with the bot is the same. If JS bot with the same structure as a C# bot acted differently from said C# bot, that's another argument for addressing the parity issue. Changing code with the intent of it acting the same way is normally a refactoring, except this scenario is not a refactoring for the first point listed.

___

TL: DR;

Fixing this issue could be breaking since production bots may be expecting the `undefined`. For bot-to-user interactions, C# and JS bots behave the same, which means the only recognizable difference in behavior is in the stored state. The bot-to-user interaction should _not_ change, so changing the code doesn't confer great benefits for bot developers, it just opens the door for potential future issues.

**Comment by v-kydela** [comment]
From my perspective, the risk of breaking existing bots is negligible enough in this case to justify fixing the parity issue. It also seems like good practice to always have dialogs initialize their instance state as soon as they begin. So I still think this should be fixed, but I understand why the team would decide not to fix it.

**Comment by stevengum** [comment]
> It also seems like good practice to always have dialogs initialize their instance state as soon as they begin

I definitely agree with this... Though I might say that after the Prompt successfully sends the activity on beginDialog, it should then set attemptCount to 1, not 0.

By the time it gets to continueDialog it should have already prompted the user, otherwise why is continueDialog being called?

ContinueDialog should only increment the counter if the expected response was not valid and the user was reprompted...

**Comment by v-kydela** [comment]
I'm not sure why attemptCount should be set to 1 when beginDialog is called. The user has made no attempt to respond to the prompt at that point.

**Comment by stevengum** [comment]
It should be 1 as the attemptCount indicates the _number of times the prompt has been executed_. Not the number of attempts from the user to respond to the prompt.

https://github.com/microsoft/botbuilder-js/blob/18a2191f51fd553c41dcddb53a0910c6f1b59d1b/libraries/botbuilder-dialogs/src/prompts/prompt.ts#L155-L160

**Comment by stevengum** [comment]
After speaking with @mrivera-ms, I'm closing this issue to avoid potentially breaking customers relying on the current state behavior.
